Source/Protobuf/ArtifactExtensions.cs

```diff
@@ -28,41 +28,99 @@ public static PbArtifact ToProtobuf(this IArtifact artifact)
+                }
+
+                error = null;
+                return true;
             }
             catch (MissingMethodException)
             {
-                throw new ArtifactIdTypeDoesNotHaveDefaultParameterlessConstructor(typeof(TArtifact), artifact, idType);
+                error = new ArtifactIdTypeDoesNotHaveDefaultParameterlessConstructor(typeof(TArtifact), source, idType);
+                return false;
             }
+            catch (Exception ex)
+            {
+                error = new CouldNotConvertProtobufArtifact(typeof(TArtifact), source, ex.Message);
+                return false;
+            }
+        }
 
-            Generation generation = artifact.Generation;
-            var constructor = typeof(TArtifact).GetConstructor(new[] { idType, generationType });
-            if (constructor == default) throw new ArtifactTypeDoesNotHaveConstructorWithIdAndGeneration(typeof(TArtifact), artifact);
+        /// <summary>
+        /// Convert a <see cref="PbArtifact"/> to a <typeparamref name="TArtifact"/>.
+        /// </summary>
+        /// <remarks>
+        /// The method assumes that the <see cref="IArtifact" /> class has a constructor that takes in
+        /// <see cref="ArtifactId" /> as first parameter and <see cref="Generation" /> as second parameter.
+        /// It also assumes that the <see cref="ArtifactId" /> type in the <see cref="IArtifact" /> only has the default parameterless constructor.
+        /// </remarks>
+        /// <param name="source"><see cref="PbArtifact"/> to convert.</param>
+        /// <typeparam name="TArtifact">The <see cref="Type" /> of the <see cref="IArtifact" />.</typeparam>
+        /// <returns>The converted <see cref="Artifact{TId}"/>.</returns>
+        public static TArtifact To<TArtifact>(this PbArtifact source)
+            where TArtifact : class, IArtifact
+            => source.TryTo<TArtifact>(out var artifact, out var error) ? artifact : throw error;
+
+        /// <summary>
+        /// Convert a <see cref="PbArtifact"/> to a tuple with an <see cref="ArtifactId" /> and a <see cref="Generation" />.
+        /// </summary>
+        /// <param name="source"><see cref="PbArtifact"/> to convert.</param>
+        /// <param name="artifact">When the method returns, a tuple with <see cref="ArtifactId" /> and <see cref="Generation" /> if conversion was successful, otherwise null.</param>
+        /// <param name="error">When the method returns, null if the conversion was successful, otherwise the error that caused the failure.</param>
+        /// <returns>A value indicating whether or not the conversion was successful.</returns>
+        public static bool TryToArtifact(this PbArtifact source, out (ArtifactId Id, Generation Generation) artifact, out Exception error)
+        {
+            artifact = default;
+            if (!source.Id.TryTo<ArtifactId>(out var id, out error))
+            {
+                return false;
+            }
```

Comment:
As I've mentioned above, I like the return something on the same line. Up to you to decide :) 
```suggestion
            if (!source.Id.TryTo<ArtifactId>(out var id, out error)) return false;
```

Reply:
Done